tests/benchmarks/test_model_schema_generation.py

```diff
@@ -219,3 +238,104 @@ def serialize_model(self) -> Any:
             return self.field
 
     benchmark(rebuild_model, ModelWithValidator)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_str_discriminator_schema_generation(benchmark):
+    class Cat(BaseModel):
+        pet_type: Literal['cat']
+        meows: int
+
+    class Dog(BaseModel):
+        pet_type: Literal['dog']
+        barks: float
+
+    class Lizard(BaseModel):
+        pet_type: Literal['reptile', 'lizard']
+        scales: bool
+
+    class Model(BaseModel):
+        pet: Union[Cat, Dog, Lizard] = Field(..., discriminator='pet_type')
+        n: int
+
+    benchmark(rebuild_model, Model)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_callable_discriminator_schema_generation(benchmark):
+    class Pie(BaseModel):
+        time_to_cook: int
+        num_ingredients: int
+
+    class ApplePie(Pie):
+        fruit: Literal['apple'] = 'apple'
+
+    class PumpkinPie(Pie):
+        filling: Literal['pumpkin'] = 'pumpkin'
+
+    def get_discriminator_value(v: Any) -> str:
+        if isinstance(v, dict):
+            return v.get('fruit', v.get('filling'))
+        return getattr(v, 'fruit', getattr(v, 'filling', None))
+
+    class ThanksgivingDinner(BaseModel):
+        dessert: Annotated[
+            Union[
+                Annotated[ApplePie, Tag('apple')],
+                Annotated[PumpkinPie, Tag('pumpkin')],
+            ],
+            Discriminator(get_discriminator_value),
+        ]
+
+    benchmark(rebuild_model, ThanksgivingDinner)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_tagged_union_with_callable_discriminator_nested_schema_generation(benchmark):
+    class BlackCat(BaseModel):
+        pet_type: Literal['cat']
+        color: Literal['black']
+        black_name: str
+
+    class WhiteCat(BaseModel):
+        pet_type: Literal['cat']
+        color: Literal['white']
+        white_name: str
+
+    Cat = Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]
+
+    class Dog(BaseModel):
+        pet_type: Literal['dog']
+        name: str
+
+    class Model(DeferredModel):
+        pet: Annotated[Union[Cat, Dog], Field(discriminator='pet_type')]
+        n: int
+
+    benchmark(rebuild_model, Model)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_untagged_union_left_to_right_mode_schema_generation(benchmark):
+    class User(DeferredModel):
+        id: Union[str, int] = Field(union_mode='left_to_right')
+
+    benchmark(rebuild_model, User)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_untagged_union_smart_mode_schema_generation(benchmark):
+    class User(DeferredModel):
+        id: Union[int, str, UUID]
+        name: str
+
+    benchmark(rebuild_model, User)
```

Comment:
These two can be removed, `union_mode` is only relevant during model validation, not schema building.